### ISSUE TITLE
Killed 9: with custom dataset

### ISSUE THREAD

**Issue by anfoss** [stellargraph/stellargraph#1839]
### Describe the bug

I have a network with 

- adjacency matrix of shape (X,X) 3 gb approx
- feature matrix of shape (X,Y) 38 mb
- binary classes

I am loading the data into StellarGraph using

        def conv2list(adj_m):
            """
            converts adjacency matrix to adj list to load into stellargraph
            """
            idx = np.triu_indices(n=adj_m.shape[0])
            v = adj_m[idx].reshape(-1, 1)
            col = idx[0].reshape(-1, 1)
            row = idx[1].reshape(-1, 1)
            final = np.concatenate((row, col, v), axis=1)
            d = pd.DataFrame(final)
            d.columns = ['source', 'target', 'weight']
            return d

         def get_data(path):
            df = pd.read_csv(path, sep="\t")
            df.fillna(0, inplace=True)
            Y = df['class'] + df['class_x'] + df['class_y']
            #  Y[np.where(Y>0)] = 1
            Y.replace(np.min(Y), 0, inplace=True)
            Y.replace(np.max(Y), 1, inplace=True)
            ids = df[['GeneNames', 'Uniprot']]
            X = df.drop(['class', 'GeneNames', 'Uniprot', 'class_x', 'class_y'], axis=1).values
            return X

    X = get_data('df_sigm.txt')

    A = np.load('adj_str.npy')

    y = pd.read_csv('classes.txt', sep="\t")['0']
    train_subjects, test_subjects = model_selection.train_test_split(
        y, test_size=0.2, stratify=y
    )
    val_subjects, test_subjects = model_selection.train_test_split(
        test_subjects, test_size=0.3, stratify=test_subjects
    )

    # this are all np array
    # y = OneHotEncoder().fit(y.reshape(-1,1)).transform(y.reshape(-1,1)).toarray()
    train_targets = train_subjects.values
    val_targets = val_subjects.values
    test_targets = test_subjects.values

    all_targets = y.values

    G = StellarGraph(X, edges=conv2list(A))
    G.info()
    StellarGraph: Undirected multigraph
    Nodes: 20379, Edges: 207662010

    Node types:
      default: [20379]
        Features: float64 vector, length 128
        Edge types: default-default->default

    Edge types:
      default-default->default: [207662010]
         Weights: range=[0, 1], mean=0.0263249, std=0.1601
         Features: none

But then when I run

    generator = FullBatchNodeGenerator(G, method="gat", sparse=False)
    train_gen = generator.flow(train_subjects.index, train_targets)

    gat = GAT(
        layer_sizes=[8, 1],
        attn_heads=8,
        generator=generator,
        bias=True,
        in_dropout=0.3,
        attn_dropout=0.3,
        activations=["elu", "sigmoid"],
        normalize=None,
        saliency_map_support=True,
    )

    x_inp, predictions = gat.in_out_tensors()

    model = Model(inputs=x_inp, outputs=predictions)
    model.compile(
        optimizer=optimizers.Adam(lr=0.005),
        loss=losses.binary_crossentropy,
        weighted_metrics=["acc"],
    )

    val_gen = generator.flow(val_subjects.index, val_targets)
    N = G.number_of_nodes()
    history = model.fit(
        train_gen, validation_data=val_gen, shuffle=False, epochs=10, verbose=1
    )

Killed: 9 at first epoch

I have 32 gb of memory and can load larger networks/train on exactly the same network  in other GNN frameworks. Is there something that I am missing like loading the data in another way?  The goal would be to obtain saliency maps for each node.

**Comment by huonw** [comment]
Thanks for including such detail in your report!

Unfortunately, I suspect this is caused by the saliency map implementation requiring dense matrices (as hinted by the `sparse=False` argument to `FullBatchNodeGenerator`). In particular, the key piece enabled by `saliency_map_support=True` does some heavy weight manipulations on dense matrices at the moment:

https://github.com/stellargraph/stellargraph/blob/ec132647e5cf43ff683e3f2e72e18ac6daa98202/stellargraph/layer/graph_attention.py#L331-L341

From a quick look, it looks like that code is potentially holding many dense N x N matrices. I suspect that code was written as a direct translation of a paper, and little thought was given to performance, given there's various things that could be factored out, to reduce the (source-level) instances, something like:

```python
exp = K.exp(dense - K.max(dense, axis=1, keepdims=True))
W = exp * (
    A * (self.delta - 2 * self.non_exist_edge * self.delta + self.non_exist_edge) +
    self.non_exist_edge * (self.delta * tf.ones((N, N)) + tf.eye(N))
)
dense = W / K.sum(W, axis=1, keepdims=True)
```

(I haven't verified that's a correct refactoring, so it'll definitely need double checking.)

One way to apply this right now without needing to do a pull request would be to copy the whole `graph_attention.py` file and adjust those lines, as well as replacing `from ..mapper` with `from stellargraph.mapper import ...` and `from.misc import ...` with `from stellargraph.layer.misc import ...`.

**Comment by anfoss** [comment]
Thanks for the quick answer, will try to do local import of graph attention and see if it works. will post here on the outcome and do a pull request in case needed.

**Comment by anfoss** [comment]
Thanks for the quick answer, but still does not work. Is there any way to import the adjacency matrix without having to specificy empty edges? Might be more memory efficient if I can throw in a numpy array with only source and target for node that actually exists

**Comment by huonw** [comment]
Unfortunately no, not for saliency maps: saliency maps only work with `sparse=False` specified to `FullBatchNodeGenerator`, which forces it to use a full NxN dense matrix. Training the model without saliency map support can use `sparse=True` (which is also the default), which uses a [sparse adjacency matrix](https://www.tensorflow.org/api_docs/python/tf/sparse/SparseTensor) along the lines of what you're hoping for.

**Comment by adocherty** [comment]
Hi,  A suggestion I would have is to use [GCN saliency which supports sparse matrices](https://stellargraph.readthedocs.io/en/stable/demos/interpretability/gcn-sparse-node-link-importance.html) – this would be a good start particularly as GAT saliency is experimental. 

Alternatively, if using GAT is important, you could try training on the full dataset with `sparse=True`, save the trained weights, and then create a new GAT model on a small sub-graph containing the nodes that you are specifically interested in, and finally loading the trained weights into this new model and calculating the saliency map using this new dense model on the sub-graph of interest.  This should give the same results in the sub-graph as long as nodes k-hops away from the target node(s) where the saliency is required are included in the sub-graph (where k is the depth of the GAT layer).

If `model` is the full sparse GAT model (with `saliency_map_support=False`), and `model2` is the dense GAT model with `saliency_map_support=True` then the weights can be transferred as follows:

```
model_weights = model.get_weights()
model2.set_weights(model_weights)
```

I hope this helps!

**Comment by anfoss** [comment]
the solution with pulling the weights out worked like a charm, thanks!
